Firestore/core/src/firebase/firestore/util/async_queue.cc

```diff
@@ -0,0 +1,314 @@
+  bool done_{};
+};
+
+}  // namespace detail
+
+using detail::DelayedOperationImpl;
+
+void DelayedOperation::Cancel() {
+  if (auto live_instance = handle_.lock()) {
+    live_instance->Cancel();
+  }
+}
+
+void AsyncQueue::RemoveDelayedOperation(const DelayedOperationImpl& dequeued) {
+  const auto new_end = std::remove_if(
+      operations_.begin(), operations_.end(),
+      [&dequeued](const OperationPtr& op) { return op.get() == &dequeued; });
+  FIREBASE_ASSERT_MESSAGE(new_end != operations_.end(),
+                          "Delayed operation not found");
+  operations_.erase(new_end, operations_.end());
+}
+
+void AsyncQueue::VerifyIsCurrentQueue() const {
+  FIREBASE_ASSERT_MESSAGE(
+      OnTargetQueue(),
+      "We are running on the wrong dispatch queue. Expected '%s' Actual: '%s'",
+      GetTargetQueueLabel().data(), GetCurrentQueueLabel().data());
+  FIREBASE_ASSERT_MESSAGE(
+      is_operation_in_progress_,
+      "VerifyIsCurrentQueue called outside enterCheckedOperation on queue '%s'",
+      GetCurrentQueueLabel().data());
+}
+
+void AsyncQueue::EnterCheckedOperation(const Operation& operation) {
+  FIREBASE_ASSERT_MESSAGE(!is_operation_in_progress_,
+                          "EnterCheckedOperation may not be called when an "
+                          "operation is in progress");
+
+  is_operation_in_progress_ = true;
+
+  VerifyIsCurrentQueue();
+  operation();
+
+  is_operation_in_progress_ = false;
+}
+
+void AsyncQueue::Enqueue(const Operation& operation) {
+  FIREBASE_ASSERT_MESSAGE(!is_operation_in_progress_ || !OnTargetQueue(),
+                          "Enqueue called when we are already running on "
+                          "target dispatch queue '%s'",
+                          GetTargetQueueLabel().data());
+  // Note: can't move operation into lambda until C++14.
+  DispatchAsync(dispatch_queue(),
+                [this, operation] { EnterCheckedOperation(operation); });
+}
+
+void AsyncQueue::EnqueueAllowingSameQueue(const Operation& operation) {
+  // Note: can't move operation into lambda until C++14.
+  DispatchAsync(dispatch_queue(),
+                [this, operation] { EnterCheckedOperation(operation); });
+}
+
+DelayedOperation AsyncQueue::EnqueueAfterDelay(const Milliseconds delay,
+                                               const TimerId timer_id,
+                                               Operation operation) {
+  // While not necessarily harmful, we currently don't expect to have multiple
+  // callbacks with the same timer_id in the queue, so defensively reject them.
+  FIREBASE_ASSERT_MESSAGE(!ContainsDelayedOperation(timer_id),
+                          "Attempted to schedule multiple callbacks with id %u",
+                          timer_id);
+
+  operations_.emplace_back(std::make_shared<DelayedOperationImpl>(
+      this, timer_id, delay, std::move(operation)));
+  return DelayedOperation{operations_.back()};
+}
+
+void AsyncQueue::RunSync(const Operation& operation) {
```

Comment:
Is this implementation identical to `AsyncQueue::Enqueue`? If logically the same, might let one call the other or just keep one.

Reply:
That's right. I was mentally auto-corrected sync to async for the `AsyncQueue` class and now realized this meant to run it syncly.